File: data_structure_python/9010.py
import sys

class Edge:
    def __init__(self,u,v,weight):
        self.u = u
        self.v = v
        self.weight = weight

#定义一个并查集
parent=[]

#实现查找和路径压缩的函数
def find(x):
    if parent[x]==x:
        return x
    else:
        parent[x]=find(parent[x])
        return parent[x]

#定义合并函数
def union(x,y):
    root_x=find(x)
    root_y=find(y)
    
    if root_x!=root_y:
        parent[root_x]=root_y
        return True
    return False

def kruskal(n,edges):
    #声明将要使用全局的parent的列表
    global parent
    
    edges.sort(key = lambda edge:edge.weight)#表明排序的键是edges的weight作为edge
    
    #先初始化独立的parent列表
    parent=[i for i in range(n+1)]
    
    total_weight =0
    edge_count = 0
    
    for edge in edges:
        if find(edge.u) != find(edge.v):
            union(edge.u, edge.v)
            total_weight += edge.weight
            edge_count += 1
            
        # 步骤 6: 优化 (当边数达到 n-1 时停止)
        if edge_count == n - 1:
            break
    return total_weight

def main():
    try:
        T=int(sys.stdin.readline())
    except ValueError:
        T=0
    for _ in range(T):
        # C++: int n, m; std::cin >> n >> m;
        try:
            line = sys.stdin.readline().split()
            if not line:
                continue # 跳过空行
            n, m = map(int, line)
        except (ValueError, IndexError):
            continue # 处理格式错误的行

        # C++: std::vector<Edge> edges(m); ...
        edges = []
        for _ in range(m):
            try:
                u, v, weight = map(int, sys.stdin.readline().split())
                edges.append(Edge(u, v, weight))
            except (ValueError, IndexError):
                continue # 处理格式错误的行

        # C++: long long mst_weight = Kruskal(n, edges);
        mst_weight = kruskal(n, edges)
        
        # C++: std::cout << mst_weight << std::endl;
        print(mst_weight)
if __name__ == '__main__':
    main()

File: data_structure_cpp/9010.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include<numeric>

struct Edge {
    int u, v;
    int weight;
    //重载<,来让sort函数可以排序
    bool operator<(const Edge& e) const {
        return weight < e.weight;
	}
};

//定义父节点
//定义并查集的数据
//parent[i]表示i的父节点s
std::vector<int> parent;

//并查集的find操作
//使用路径压缩优化
int find(int i) {
    if (parent[i] == i) {
    return i;
    }
    return parent[i] = find(parent[i]);
}

//并查集的合并操作
bool unite(int i, int j) {
  int root_i = find(i);
  int root_j = find(j);

  if (root_i != root_j) {
    //将一个根连接到另一个根上
      parent[root_i] = root_j;
      return true;
  }
  return false;
}
long long  Kruskal(int n,std::vector<Edge>& edges) {
    //首先将边按照权重排序
    std::sort(edges.begin(), edges.end());

    //初始化并查集
    //需要n+1个空间，顶点是从1~n
    parent.resize(n + 1);
    //初始化的时候每个节点都是自己的父节点
    for (int i = 0; i <= n; ++i) {
		parent[i] = i;
    }
    long long total_weight = 0;
    int edge_count = 0;

    for (const auto& edge : edges) {
        if (find(edge.u) != find(edge.v)) {
        //如果这两条边是不联通的
            unite(edge.u, edge.v);
            total_weight += edge.weight;
            edge_count++;
        }
        if (edge_count == n - 1) {
            break;
        }
    
    }
    return total_weight;
    
}
int main() { 
    int T;
	std::cin >> T;
    while (T--) {
        int n, m;
        std::cin >> n >> m;

        // 存储所有 m 条边
        std::vector<Edge> edges(m);
        for (int i = 0; i < m; ++i) {
            std::cin >> edges[i].u >> edges[i].v >> edges[i].weight;
		}


        //开始kruskal算法
        long long mst_weight = Kruskal(n, edges);

        // 输出结果
        std::cout << mst_weight << std::endl;
      
    }
    return 0;
}